### PR TITLE
Type-only imports and exports

### LINKED ISSUE #9191: Should be possible to force import of "unused" modules in case there are side-effects
**TypeScript Version:** 

1.8.10

**Code**

_MyService.ts_

``` ts
// A self-contained demonstration of the problem follows...
import {ngModule} from './../module.ts';

export class MyService {
}

ngModule.service('myService', MyService); // <- REALLY IMPORTANT SIDE-EFFECT
```

_MyComponent.ts_

``` ts
import {MyService} from './MyService.ts';

class MyComponent {
  constructor(private myService: MyService) { }
  // ...
}
```

**Expected behavior:**
(I know this sort-of-tree-shaking is "by design", but by the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), expected behavior is this)

_MyComponent.js_

``` js
var MyService_1 = require('./MyService');
var MyComponent = (function() {
  function MyComponent(myService) {
    this.myService = myService;
  };
  // ...
})();
```

**Actual behavior:** 
_MyComponent.js_

``` js
var MyComponent = (function() {
  function MyComponent(myService) {
    this.myService = myService;
  };
  // ...
})();
```

And then AngularJS is all "myServiceProvider not found" and crashy.

### LINKED ISSUE #34750: Add syntax hint to allow reexports of types in isolatedModules mode
## Search Terms

isolatedModules, type export, type re-export, babel

## Context

In `--isolatedModules` mode you get an error when reexporting types, the reason why this breaks is documented in the [new handbook](https://microsoft.github.io/TypeScript-New-Handbook/reference/compiler-options/#exports-of-non-value-identifiers). In the following example lines 2 and 5 both raise give "Cannot re-export a type when the '--isolatedModules' flag is provided." warnings.

```ts
import { SomeType } from "someModule";
export { SomeType };
let x: SomeType | undefined;

export { SomeOtherType } from "someModule";
```

This leaves programme authors who wish to compile using a per-file compiler with some choices: 

Use `export * from "someModule"`, which will deal with all exports, but won't work if you need to cherry pick exports. It's usually possibly to rearchitect your folder structure to work around this but not always.

Or use a temporary renaming:
```ts
import { SomeType as SomeType1, SomeOtherType as SomeOtherType1 } from "someModule";
export type SomeType = SomeType1;
let x: SomeType | undefined;

export type SomeOtherType = SomeOtherType1;
```

Prior to #31231 (to be released in TS 3.7) authors could export a type that matches the same name as an import, but that loophole has been closed:

```ts
import { SomeType, SomeOtherType } from "someModule";
export type SomeType = SomeType;
let x: SomeType | undefined;

export type SomeOtherType = SomeOtherType;
```

Authors should not have to go through such renaming dances in order to get type reexports working in isolatedModules mode, instead it should be possible to hint that an import / reexport is of a type.

## Suggestion

Add some kind of syntax hint to imports and reexports that denote that a particular identifier is a type and thus can be elided when compiling in isolatedModules mode.

Here I'm proposing the strawman of a `type` keyword before the ImportSpecifier / ExportSpecifier, as suggested by @Jessidhia in https://github.com/microsoft/TypeScript/pull/31231#issuecomment-537383472. This is the same syntax [Flow uses](https://flow.org/en/docs/types/modules/) though I must admit I'm not familiar with Flow.

```ts
import { type SomeType } from "someModule";
export { SomeType };
let x: SomeType | undefined;

export { type SomeOtherType } from "someModule";
```

Placing the hint on a per-specifier basis will allow mixing of type and value import/reexports in a single import/export statement:

```ts
import { SomeValue, type SomeType, SomeOtherValue, type SomeOtherType } from "someModule";

export { AnotherValue, type AnotherType, YetAnotherValue, type YetAnotherType } from "someOtherModule";
```

Regarding unhappy paths:

- Adding a `type` hint to a value should raise a type error "Value export must not be tagged as a type"
- When isolatedModules is enabled type exports without a type hint should continue to raise a type error, though the message will likely need to be updated to something like "Cannot rexport type without a `type` hint"
- When isolatedModules is disabled then there is no value to adding the `type` hint. We could disable adding the hints by throwing a TypeError "Type exports with `type` hint are only required in isolatedModules mode" but that feels pretty heavy handed. I think a better path is allowing the type hint from the compiler's perspective, while having a linting rule to that can can say if these type hints should be present or not (defaulting to present for isolatedModules is enabled and absent if disabled). This will help projects who wish to migrate from using `tsc` to a per-file compiler.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by andrewbranch**
**TL;DR**:
- `import type { A } from './mod'`, `export type { A } from './mod'`
- Add flag to stop eliding import declarations from emit

**To do:**
- [x] parsing
- [x] checking: default import
- [x] checking: named imports
- [x] checking: named exports
- [x] checking: namespace import
- [x] update isolatedModules error message
- [x] test with enums
- [x] isolatedModules code fix
- [x] auto imports behavior
- [x] compiler flag, emit behavior
- [x] checker API (`getTypeAtLocation`)
- [x] grammar error for mixing default and named bindings
- [x] code fix for splitting a default + named bindings into two import declarations
- [x] error for usage in JS
- [x] check/improve parsing diagnostics for common mistakes
- [x] test more quick info, completions, rename, etc.
- [x] TextMate grammar
- [ ] auto-import/codefix for name in value space that’s already type-only imported?
- [ ] code fix for `--importsNotUsedAsValue=error` error

## Background

TypeScript elides import declarations from emit where, in the source, an import clause exists but all imports are used only in a type position [[playground]](https://www.typescriptlang.org/play/#code/JYWwDg9gTgLgBAbzgeQHYBsCeBVAzgUwBMBBXAFUzHzgF84AzKCEOAIlYG4AoUSWROHiKkAagEN0AV2p1GzNpy5dC+AMboxUaqoipc8MQC4UGHARLlK+bjr3wARnAC8g86InTuQA). This sometimes creates confusion and frustration for users who write side-effects into their modules, as the side effects won’t be run if other modules import only types from the side-effect-containing module (#9191).

At the same time, users who transpile their code file by file (as in Babel, ts-loader in `transpileOnly` mode) sometimes have the _opposite_ problem, where a re-export of a type _should_ be elided, but the compiler can’t tell that the re-export is only a type during single-file transpilation (#34750, https://github.com/TypeStrong/ts-loader/issues/751) [[playground]](https://www.typescriptlang.org/play/#code/PTAEHUEsBcAsHsCu1QCMCmkB2BzUBDVAG3VGnlCPngGsCUADAOmHyegGcGAaUAd3QAoEKAAm8LAHIUNLPD79Y6OOgBOZWJA6gtBMgE8ADqXjr8oAG74iiE+tTw43YWA4UBodAFsYO6L1RkP35IIiI0UnwsT1VVU3pQVUQsaEgvdCZBdAAPQ1MUAG9QAClEDmgAQQAVI1IAX1AAMzivUEkWfEkAbiA).

## Prior art

In early 2015, [Flow introduced type-only imports](https://flow.org/blog/2015/02/18/Import-Types/) which would not be emitted to JS. (Their default behavior, in contrast to TypeScript’s, was never to elide imports, so type-only imports for them were intended to help users cut down on bundle size by removing unused imports at runtime.)

Two months later, #2812 proposed a similar syntax and similar emit behavior for TypeScript: the compiler would stop eliding import declarations from emit unless those imports were explicitly marked as type-only. This would give users who needed their imports preserved for side effects exactly what they wanted, and also give single-file transpilation users a syntactic hint to indicate that a re-export was type-only and could be elided: `export type { T } from './mod'` would re-export the type `T`, but have no effect on the JavaScript emit.

#2812 was ultimately declined in favor of introducing the `--isolatedModules` flag, under which re-exporting a type is an error, allowing single-file transpilation users to catch ambiguities at compile time and write them a different way.

## Since then

Over the last four years after #2812 was declined, TypeScript users wanting side effects have been consistently confused and/or frustrated. They have workarounds (read #9191 in full for tons of background and discussion), but they’re unappealing to most people.

For single-file transpilation users, though, two recent events have made their lives harder:

<ol>
<li>

In TypeScript 3.7, we sort of took away `--isolatedModules` users’ best workaround for reexporting a type in #31231. Previously, you could replace `export { JustAType } from './a'` with

```ts
import { JustAType } from './a';
export type JustAType = JustAType;
```

But as of TypeScript 3.7, we disallow the name collision of the locally declared `JustAType` with the imported name `JustAType`.

</li>
<li>

If a Webpack user was left with an erroneous `export { JustAType } from './a'` in their output JavaScript, Webpack 4 would warn, but compilation would succeed. Many users simply ignored this warning (or even filtered it out of Webpack’s output). But in Webpack 5 beta, @sokra has expressed some desire to [make these warnings _errors_](https://github.com/webpack/webpack/issues/9802#issuecomment-547328259).

</li>
</ol>

## Proposal

- Add type-only imports and exports similar to #2812 and Flow
- ~~Change the default emit behavior of the compiler to stop eliding regular imports even if the imported names are only used in type positions~~
- Always elide imports and re-exports explicitly marked as type-only
- ~~Add a (temporary?) compiler flag that restores the current behavior of eliding imports that are used only for types to help users with back-compat~~
- **Updated**: Leave the default emit as-is, adding the flag `--importsNotUsedAsValue <"remove" | "preserve" | "error">` to control the behavior
  - `remove` is default; maintains today’s behavior
  - `preserve` keeps imports used only for types in the emit as a side-effect import
  - `error` acts as `preserve` but also adds an error whenever an `import` could be written as an `import type`

### Syntax

Supported forms are:

```ts
import type T from './mod';
import type { A, B } from './mod';
import type * as Types from './mod';

export type { T };
export type { T } from './mod';
```

Possible additions but I think not terribly important:

```ts
export type * from './mod';
export type * as Types from './mod'; // pending #4813
```

We notably do _not_ plan to support at this time:

- `type` modifier on import/export specifiers: `import { type A } from './mod'`, `export { A, type B }`
- Mixing a type-only default import with named or namespace imports: `import type T, { A } from './mod'`, `import type T, * as ns from './mod'`

The forms in the former bullet will be syntax errors; the forms in the latter will be grammar errors. We want to start with productions that can be read unambiguously, and it’s not immediately clear (especially in the absence of Flow’s implementation), what the semantics of `import type A, { B } from './mod'` should be. Does `type` apply only to the default import `A`, or to the whole import clause? We prefer no one need wonder.

### Type semantics

Any symbol with a type side may be imported or exported as type-only. If that symbol has no value side (i.e., is _only_ a type), name resolution for that symbol is unaffected. If the symbol _does_ have a value side, name resolution for that symbol will see _only_ the type side. The typical example is a class:

```ts
// @Filename: /a.ts
export default class A {}

// @Filename: /b.ts
import type A from './a';
new A();
//  ^ 'A' only refers to a type, but is being used as a value here.

function f(obj: A) {} // ok
```

If the symbol is a namespace, resolution will see a mirror of that namespace recursively filtered down to just its types and namespaces:

```ts
// @Filename: /ns.ts
namespace ns {
  export type Type = string;
  export class Class {}
  export const Value = "";
  export namespace nested {
    export class NestedClass {}
  }
}
export default ns;

// @Filename: /index.ts
import type ns from './ns';
const x = ns.Value;
//        ^^ Cannot use namespace 'ns' as a value.

let c: ns.nested.NestedClass;
```

### Emit

_**Updated**: When the `importsNotUsedAsValue` flag is set to 'preserve',_ type-only import declarations will be elided. Regular imports where all imports are unused or used only for types will _not_ be elided (only the import clause will be elided):

```ts
// @importsNotUsedAsValue: preserve

// @Filename: /a.ts
import { T } from './mod';
let x: T;

// @Filename: /a.js
import "./mod";
let x;
```

### Back-compat ~~flag~~

~~There’s a new flag `removeUnusedImports`. Its name is not perfect because it really means “remove imports that have imported names that never get used in a value position.” Open to suggestions.~~

**Updated**: this PR is backward-compatible by default.

### Auto-imports behavior

- Symbols without a value side will be imported as type-only if there’s not already a regular import from the containing module. If there’s an existing import from the containing module, it will be added to that import (as happens today).
- There will be a configuration option to disable type-only auto imports entirely (since some people use VS Code’s TypeScript version for editor services but compile with an older version).

I’m not yet confident what other changes, if any, will the right move, but the main scenarios to consider are:

- User auto imports a class, enum, or namespace in a type position. Should we do a type-only import?
- User has a type-only import of a class, enum, or namespace, then later tries to use the same symbol in a value position. Do we convert the type-only import to a regular import? (Is that even possible with a completions code action?)

---

Successor of #2812
Closes #9191
Closes #34750 

Would close if they were still open:
- #20433
- #16472
- #18586
- #30525
- #28285

### CODE REVIEW

**Comment by andrewbranch** [comment]
Meant to open this as a draft PR. Implementation is unfinished.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2332,17 +2341,81 @@ namespace ts {
...
+                return createNamespaceModuleForModule(target);
+            }
+            if (!(target.flags & SymbolFlags.Value)) {
+                return target;
+            }
+            if (target.flags & SymbolFlags.Type) {
+                const alias = createSymbol(SymbolFlags.TypeAlias, target.escapedName);
+                alias.declarations = emptyArray;
+                alias.immediateTarget = target;
+                return alias;
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: As the implementation has progressed, I’ve had to diverge handling of these symbols from normal handling of type aliases in a couple places, so it now feels like a flagrant abuse of `SymbolFlags.TypeAlias`. But I notice we’re running out of room on `SymbolFlags`. Any suggestions?

**Comment by ajafff** [comment]
@andrewbranch what about imported values that are only used for their types via `typeof` in the file? Are these imports still elided?

**Comment by andrewbranch** [comment]
@ajafff I think ideally the plan would be no, imports not marked with `type` are never elided. Flow has an `import typeof` form for this use case. I think that’s probably a reasonable follow-up feature. I had initially thought of `import typeof` as syntactic sugar for something already possible, but as you bring up, if you care about eliding imports that are unnecessary at runtime but you need the `typeof` a value, the original proposal doesn’t allow for that. /cc @DanielRosenwasser thoughts?

**Comment by andrewbranch** [comment]
Updated the PR description with changes from the decisions made in the last two design meetings (#35806, #35807). Summary:

1. **Problem**: concern that a change to default emit behavior will break people’s apps in subtle, hard-to-diagnose ways. **Solution**: don’t change the default emit. Opt-in with a compiler flag.
2. **Problem**: if I want to opt into the new emit behavior and use a type-only import everywhere that I can, how do I track down my existing imports that should become `import type`? **Solution**: make the compiler flag have a setting where import declarations that only get used as types are errors, forcing you to use `import type` where possible.

Current status: 1 is done. The flag is now called `importsNotUsedAsValue`, and its value is currently of type `"remove" | "preserve"`, with default `"remove"`. Using a string value leaves room for a third option, `"error"`, to fulfill Problem 2.

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Added a couple of changes to phrasing. I still am not a fan of the option name, but I think we'd like to at least get this in for the beta.

**Comment by DanielRosenwasser** [comment]
@SeaRyanC if I don't get the chance to in the next 30, feel free to merge this when the tests pass.

**Comment by johnnyreilly** [comment]
I just noticed this PR and I wanted to say I think this is a *great* addition!

I wanted to share a use case that's particularly close to my heart; you may be aware of [fork-ts-checker-webpack-plugin](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin). Amongst other things, it powers the TypeScript checking experience in create-react-app.

I worked on amending the plugin so that it could be shipped with create-react-app. Facebook had a hard requirement that the plugin not have a dependency upon TypeScript. We were able to support this by rewriting the plugin somewhat. However, we still needed a bunch of types at compile time. As a consequence there's a whole bunch of imports followed by `// Imported for types alone` comments:

https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/blob/f07881e80e522c6233568e9f5f7b9574dbae750c/src/ApiIncrementalChecker.ts#L2

I got this wrong a bunch of times a created unwitting dependencies. As I understand this feature, I think `"importsNotUsedAsValue": "error"` would have made that journey easier and could prevent future regressions. Rejoice!

**Comment by nicolo-ribaudo** [comment]
:heart::heart: from Babel for this PR!

**Comment by stevefan1999-personal** [comment]
Thank you sir

**Comment by johnnyreilly** [comment]
One more thought; related to this ts-loader issue: https://github.com/TypeStrong/ts-loader/issues/751

I wonder if we should consider changing the default ts-loader `tsconfig` compiler flag to:

`--importsNotUsedAsValue "preserve"`

For webpack 5 given that:

> preserve keeps imports used only for types in the emit as a side-effect import

I think this might resolve the issue reported in the linked issue so most users just don't bump into it.

We'd look to do this as a breaking changes version if we did.

A little background: as policy we deliberately don't change the default compiler flags often in ts-loader. But this would seem to be a case which is very webpack specific and consequently appropriate. 

I'd be interested in your thoughts @andrewbranch ?

**Comment by elektronik2k5** [comment]
This is an awesome and long awaited feature! :champagne: 

@johnnyreilly and anyone else who needs this feature today without waiting for `typescript@3.8`: type only imports (albeit with a confusing and awkward syntax) have been part of the language [since `typescript@2.9`](https://devblogs.microsoft.com/typescript/announcing-typescript-2-9-2/), for almost two years now.

```ts
type SomeType = import('path-or-package').SomeExportedType
// or inline:
function(foo: import('path-or-package').SomeExportedType) { /* function body */ }
```

**Comment by johnnyreilly** [comment]
@elektronik2k5 I saw your thread on twitter! Thanks for sharing; somehow that had passed me by!

**Comment by ExE-Boss** [comment]
Is&nbsp;it&nbsp;possible to&nbsp;import&nbsp;types from&nbsp;`@types/*`&nbsp;packages&nbsp;directly?

## Example:
```ts
import type Foo from "@types/foo";
import type { Bar } from "@types/bar";
```

---

This&nbsp;would be&nbsp;particularly&nbsp;useful for&nbsp;importing&nbsp;types from&nbsp;non‑**npm**&nbsp;`@types/*`&nbsp;packages.

**Comment by andrewbranch** [comment]
@ExE-Boss can you not already do this by simply leaving off the `@types/` prefix? https://www.typescriptlang.org/play/#code/JYWwDg9gTgLgBAbzgQRhEBRAdgN2FCLEAUy3gF84AzAkOAcgEM0R6BuAWAChuqBXLAGMYwQnAAmEAMrpiMABbAsAcwAUpHAC4ULbHgJFSMAJSJucOBoB0AI2LEwq45y7luQA

**Comment by ExE-Boss** [comment]
Well, I&nbsp;mean, you&nbsp;can, but&nbsp;it&nbsp;feels a&nbsp;bit&nbsp;weird to&nbsp;import&nbsp;types from&nbsp;a&nbsp;package that&nbsp;doesn’t provide&nbsp;any&nbsp;exports (and&nbsp;might&nbsp;not&nbsp;even exist&nbsp;at&nbsp;runtime).

**Comment by andrewbranch** [comment]
I think adding the `type` keyword would make me more comfortable with that, but I agree it feels a little strange with the non-npm package scenario—I hadn’t considered that until you brought it up. Feel free to open a new issue to discuss.

**Comment by tmporterSGSI** [comment]
Is there a plan to support named type exports? 

For example,
```ts
export type {A as B} from "./A";
```
currently results in a syntax error (typescript version 3.8.2).

This is particularly useful in situations where you are re-exporting types from multiple modules using the 
```ts
export * from "./A";
export * from "./B";
```
syntax and both modules export types with the same name.

**Comment by fwienber** [comment]
Maybe I'm missing something and I'm not sure where to post this remark, but should `import type` really make the importing `ts` file a module?
My use case: I implement a class in TypeScript as a module, but want to use its interface in a non-module script. Currently, I have to extract the interface into a non-module, and all its transitive dependencies, too, otherwise my non-module becomes a module just because it's importing a type.
Alternative workaround: use `import(<module>).default` in the non-module _everywhere_ the type is used (I cannot use a local type declaration in global scope, as that would be visible everywhere).
I thought the definition of when a `js` file becomes a module is whenever it imports or exports anything. Type-only import/export, which is eliminated in JavaScript output, should not influence that behavior.

**Comment by andrewbranch** [comment]
I think that’s a reasonable question, and I did think about it while writing this feature. I’ll try to explain my thought process.

> I thought the definition of when a js file becomes a module is whenever it imports or exports anything. Type-only import/export, which is eliminated in JavaScript output, should not influence that behavior.

So first, just to clear up the background and definitions—it sounds like you probably know this, but just making sure the grounding for the rest of my argument is established. When we’re talking about JavaScript you’re correct: it is true that an import or export makes a file a module. However, the converse is not true. A file without an import or export is both a valid module and a valid script. But for the purposes of type checking, we have to make a decision about how to treat every file. So if we don’t see any imports or exports, we unequivocally treat the file as a script—it’s easy and harmless to add `export {}` if you _don’t_ want the file to be a script.

Now, consider that we’ve always elided unused imports from our JS emit. (As of this PR, that’s configurable, but the default is still to elide unused imports.) If you write a TypeScript file that looks obviously module-like because you `import`ed something, but then never used it in a value position, the [output JS](https://www.typescriptlang.org/play/#code/JYWwDg9gTgLgBAbzgYygUwIYzQUQDZohoB28AvnAGZQQhwDk6GyM9A3AFAfITEDOEAgDo8EAOYAKAMwBKTkA) will look like a valid script. Does that mean we should treat the TypeScript file as a script? I don’t think so. The source file _must_ be parsed as a module, and the output _can_ be parsed as a module, so treating the file as a module seems like the best choice.

When you write an `import type` declaration, it’s essentially the same as writing an `import` declaration that you never use in an emitting position, except that it’s _enforced_ that you never use it in an emitting position. Now, since `import type` is not standard ES grammar, I think you could reasonably argue that we could implement whatever rules we wanted; that it would not be inherently incorrect to say that `import type` does not constitute a module. But ultimately, given how similar it is to writing a regular `import` of a type, I think maintaining consistency that all import declarations are module markers was the correct choice.

That said, I 100% agree with you that referencing modules in scripts is painful—I’ve hit that exact problem before. But I don’t think `import type` would have been the best solution, even if we hadn’t already settled on the current behavior.

**Comment by johnnyreilly** [comment]
This wasn't my question, but I just wanted to give props to @andrewbranch for such a thoughtful and clear answer. Well done sir!

**Comment by fwienber** [comment]
Well, it was _my_ question, and although the outcome is maybe not what I _wanted_ to hear, I must second that this is a very thoughtful and clear answer and I agree 100%. Thank you, much appreciated!

**Comment by nyngwang** [comment]
> `preserve` keeps **imports** used only for types in the emit as a side-effect import

While it's a bit late, I still want to point out that the wording is a bit confusing. 

1. I first read the source issue #2812, and by their "Proposed change 1. Do not elide module **imports**", I thought the word "imports" referred to individual _names_ imported, e.g. `AType` in `import { AType }`, as the example code they provided.
2. This unfortunately caused a problem when I tried to understand the bolded "imports" quoted above. I believe it's not just me, as there is #43393. It took me some time to realize that "imports" means every `import`-statement for names without value-side usage. (and I indeed found the naming `importsNotUsedAsValues` precise and clean)

I hope this could help someone like me who is interested in the history of deprecated options covered by the new `verbatimModuleSyntax` option in the future.

All in all, I like TypeScript :)
